packages/databyss-api/src/routes/api/helpers/pagesHelper.js

```diff
@@ -73,3 +75,192 @@ export const composeBlockList = list => {
+    }
+    case 'blockCache': {
+      const _blockId = p.path[1]
+      const _type = typeof p.value === 'string' ? p.value : p.value.type
+
+      const _block = await Block.findOne({ _id: _blockId })
+
+      let _entityId
+      if (p.value.entityId) {
+        _entityId = p.value.entityId
+      } else {
+        // get property name from DB
+        _entityId =
+          _block[
+            {
+              ENTRY: 'entryId',
+              SOURCE: 'sourceId',
+              TOPIC: 'topicId',
+            }[_block.type]
+          ]
+      }
+
+      // set property name
+      const idType = {
+        ENTRY: { entryId: _entityId },
+        SOURCE: { sourceId: _entityId },
+        TOPIC: { topicId: _entityId },
+        AUTHOR: { authorId: _entityId },
+        LOCATION: { locationId: _entityId },
+      }[_type]
+
+      const blockFields = {
+        type: _type,
+        _id: _blockId,
+        user: req.user.id,
+        account: req.account._id,
+        ...idType,
+      }
+
+      // TODO: old idType still exists in database
+      await Block.findOneAndUpdate({ _id: _blockId }, blockFields, {
+        new: true,
+      })
+
+      break
+    }
+    case 'blocks': {
+      const _index = p.path[1]
+      // insert block id into page
+      const _page = await Page.findOne({ _id: req.page._id })
+      const blocks = _page.blocks
+      blocks.splice(_index, 1, { _id: p.value._id })
+      await Page.findOneAndUpdate({ _id: req.page._id }, { blocks })
+
+      break
+    }
+    case 'selection': {
+      const _id = p.value._id
+      if (_id) {
+        await Selection.findByIdAndUpdate({ _id }, { $set: p.value })
+      }
+
+      break
+    }
+    default:
+      break
+  }
+}
+
+const addPatch = async (p, _cache, req) => {
+  const _prop = p.path[0]
+
+  switch (_prop) {
+    case 'blocks': {
+      const _index = p.path[1]
+      // insert block id into page
+      const _page = await Page.findOne({ _id: req.page._id })
+      const blocks = _page.blocks
+      blocks.splice(_index, 0, { _id: p.value._id })
+      await Page.findOneAndUpdate({ _id: req.page._id }, { blocks })
+
+      break
+    }
+    case 'blockCache': {
+      const _type = p.value.type
+      const _entityId = p.value.entityId
+      const _blockId = p.path[1]
+
+      // add entity id to temporary cache
+      _cache[_entityId] = _blockId
+
+      const idType = {
+        ENTRY: { entryId: _entityId },
+        SOURCE: { sourceId: _entityId },
+        TOPIC: { topicId: _entityId },
+        AUTHOR: { authorId: _entityId },
+        LOCATION: { locationId: _entityId },
+      }[_type]
+
+      const blockFields = {
+        type: _type,
+        _id: _blockId,
+        user: req.user.id,
+        account: req.account._id,
+        ...idType,
+      }
+
+      const _block = new Block(blockFields)
+      await _block.save()
+      break
+    }
+    case 'entityCache': {
+      const entityFields = {
+        text: p.value.text ? p.value.text : p.value,
```

Comment:
when does it only have `value` vs `value.text`?

Reply:
immer payload may differ. sometimes the payload `value` includes a `text` property, other times, the `text` is the `value` property.